cdap-client/src/main/java/co/cask/cdap/client/AuthorizationClient.java

```diff
@@ -202,10 +202,12 @@ private HttpResponse executePrivilegeRequest(EntityId entityId, Principal princi
     FeatureDisabledException, UnauthenticatedException, IOException, NotFoundException, UnauthorizedException {
     HttpResponse httpResponse = doExecuteRequest(request, HttpURLConnection.HTTP_NOT_FOUND);
     if (HttpURLConnection.HTTP_NOT_FOUND == httpResponse.getResponseCode()) {
-      if (!listAllRoles().contains(new Role(principal.getName()))) {
-        throw new NotFoundException("Role: " + principal.getName());
+      if (principal.getType().equals(Principal.PrincipalType.ROLE) &&
+        httpResponse.getResponseMessage().equals(String.format("%s not found.", principal.getName()))) {
```

Comment:
Also it might be better to Refactor RoleNotFoundException to PrinicpalNotFoundException and then use that here instead of just NotFoundException.